File: example/cpp_lua/testmsg.cpp
#include "testmsg.h" 
#include "ByteBuffer.h" 
bool testmsg::read(ByteBuffer * buffer)
{
	if(sizeof(id) + buffer->rpos() > buffer->wpos()) return false;
	*buffer >> id;
	if(play.read(buffer) == false) return false;
	int len_array;
	if(sizeof(len_array) + buffer->rpos() > buffer->wpos()) return false;
	*buffer >> len_array;
	for( int i = 0; i< len_array; ++i) { 
		int32 temp_array;
		if(sizeof(temp_array) + buffer->rpos() > buffer->wpos()) return false;
		*buffer >> temp_array;
		array.push_back(temp_array);
	}
	int len_attrs;
	if(sizeof(len_attrs) + buffer->rpos() > buffer->wpos()) return false;
	*buffer >> len_attrs;
	for( int i = 0; i< len_attrs; ++i) { 
		Attr temp_attrs;
		if(temp_attrs.read(buffer) == false) return false;
		attrs.push_back(temp_attrs);
	}
	int len_vstr;
	if(sizeof(len_vstr) + buffer->rpos() > buffer->wpos()) return false;
	*buffer >> len_vstr;
	for( int i = 0; i< len_vstr; ++i) { 
		std::string temp_vstr;
		*buffer >> temp_vstr;
		vstr.push_back(temp_vstr);
	}
	return true;
}
void testmsg::write(ByteBuffer * buffer)
{
	*buffer << id;
	play.write(buffer);
	int len_array = array.size();
	*buffer << len_array;
	for( int i = 0; i< len_array; ++i) { 
		*buffer << array[i];
	}
	int len_attrs = attrs.size();
	*buffer << len_attrs;
	for( int i = 0; i< len_attrs; ++i) { 
		attrs[i].write(buffer);
	}
	int len_vstr = vstr.size();
	*buffer << len_vstr;
	for( int i = 0; i< len_vstr; ++i) { 
		*buffer << vstr[i];
	}
}

File: example/cpp_lua/testmsg.lua
require("types")
testmsg = {}
testmsg.__index = testmsg
function testmsg:read(buffer)
	if 4 + buffer:rpos() > buffer:wpos() then return false end
	self.id = buffer:getUint32();
	self.play = Player:new()
	if self.play:read(buffer) == false then return false end;
	local len_array = 0
	local temp_array = nil
	if 4 + buffer:rpos() > buffer:wpos() then return false end
	len_array = buffer:getInt32();
	for i = 1, len_array, 1 do
		if 4 + buffer:rpos() > buffer:wpos() then return false end
		temp_array = buffer:getInt32();
		self.array[i] = temp_array
	end
	local len_attrs = 0
	local temp_attrs = nil
	if 4 + buffer:rpos() > buffer:wpos() then return false end
	len_attrs = buffer:getInt32();
	for i = 1, len_attrs, 1 do
		temp_attrs = Attr:new()
		if temp_attrs:read(buffer) == false then return false end;
		self.attrs[i] = temp_attrs
	end
	local len_vstr = 0
	local temp_vstr = nil
	if 4 + buffer:rpos() > buffer:wpos() then return false end
	len_vstr = buffer:getInt32();
	for i = 1, len_vstr, 1 do
		temp_vstr = buffer:getString();
		self.vstr[i] = temp_vstr
	end
	return true;
end
function testmsg:write(buffer)
	buffer:pushUint32(self.id);
	self.play:write(buffer);
	local len_array = #self.array
	buffer:pushInt32(len_array);
	for i = 1, len_array, 1 do 
		buffer:pushInt32(self.array[i]);
	end
	local len_attrs = #self.attrs
	buffer:pushInt32(len_attrs);
	for i = 1, len_attrs, 1 do 
		self.attrs[i]:write(buffer);
	end
	local len_vstr = #self.vstr
	buffer:pushInt32(len_vstr);
	for i = 1, len_vstr, 1 do 
		buffer:pushString(self.vstr[i]);
	end
end
function testmsg:new()
	local o = {}
	setmetatable(o, testmsg)
	o.id = 0
	o.play = {}
	o.array = {}
	o.attrs = {}
	o.vstr = {}
	return o
end


